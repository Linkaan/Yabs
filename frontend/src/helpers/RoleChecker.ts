import UsersModule, { User } from '../store/modules/UsersModule';

export default class RoleChecker {

  private static checkPermission(user: User, required: number): boolean {
    return user && ((user.role & required) == required || (user.role > required));
  }
  
  static isPrincipal(): boolean {
    return this.checkPermission(UsersModule.currentUser, 32);
  }

  static isAdmin(): boolean {
    return this.checkPermission(UsersModule.currentUser, 16);
  }

  static isCaretaker(): boolean {
    return this.checkPermission(UsersModule.currentUser, 8);
  }

  static isTeacher(): boolean {
    return this.checkPermission(UsersModule.currentUser, 4);
  }

  static isStudentHealth(): boolean {
    return this.checkPermission(UsersModule.currentUser, 2);
  }

  static isStudent(): boolean {
    return UsersModule.currentUser && (UsersModule.currentUser.role & 1) == 1;
  }
}
