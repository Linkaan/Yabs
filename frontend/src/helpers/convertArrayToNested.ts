export default function convertList(payload: any[], typeOfId: string): object { //eslint-disable-line
  return payload.reduce((acc, item) => {
    acc[item[typeOfId]] = item;
    return acc;
  }, {});
}
