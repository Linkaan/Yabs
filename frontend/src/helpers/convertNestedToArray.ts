export default function convertList(payload: any, typeOfId: string): any[] { //eslint-disable-line
  const res = Object.keys(payload).map((id) => payload[id]);
  return res;
}
