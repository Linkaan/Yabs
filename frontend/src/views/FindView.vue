<template>
  <v-container>
    <v-row>
      <v-col>
        <v-card>
          <v-chip-group mandatory>
            <v-chip
              v-if="!RoleChecker.isStudent()"
              v-model="users"
              class="ma-2"
              color="success"
              outlined
            >
              <v-icon left>
                mdi-account-outline
              </v-icon>
              Users
            </v-chip>

            <v-chip
              v-model="loans"
              class="ma-2"
              color="deep-purple"
              outlined
            >
              <v-icon left>
                mdi-handshake
              </v-icon>
              Loans
            </v-chip>
            <v-chip
              v-model="books"
              class="ma-2"
              color="light-blue"
              outlined
            >
              <v-icon left>
                mdi-library
              </v-icon>
              Books
            </v-chip>
          </v-chip-group>

          <LoanListComponent
            v-if="loans"
            :items-per-page="15"
          />
          <BookListComponent
            v-if="books"
            :use-actions="true"
            :items-per-page="15"
          />
          <UserListComponent
            v-if="users"
            :items-per-page="15"
          />
        </v-card>
      </v-col>
    </v-row>
  </v-container>
</template>

<script lang="ts">
import {Vue, Component, Prop} from 'vue-property-decorator';

import BookListComponent from '@/components/BookListComponent.vue';
import UserListComponent from '@/components/UserListComponent.vue';
import LoanListComponent from '@/components/LoanListComponent.vue';
import BooksModule from '@/store/modules/BooksModule';
import UsersModule from '@/store/modules/UsersModule';
import LoansModule from '@/store/modules/LoansModule';
import RoleChecker from '../helpers/RoleChecker';

@Component({
  components: {
    BookListComponent,
    UserListComponent,
    LoanListComponent
  },
})
export default class FindView extends Vue {
  private RoleChecker: RoleChecker = RoleChecker;
  users:boolean = false;
  loans:boolean = false;
  books:boolean = false;

  private created(): void {
    BooksModule.fetchAll();
    UsersModule.fetchAll();
    LoansModule.fetchAll();
  }

}
</script>
