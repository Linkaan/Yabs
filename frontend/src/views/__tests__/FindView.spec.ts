import { shallowFactory } from '@/helpers/testFactoryHelpers';
import { Wrapper } from '@vue/test-utils';
import FindView from '../FindView.vue';
import Vue from 'vue';

describe('FindView.vue', () => {
  it('is not empty', () => {
    const wrapper: Wrapper<Vue> = shallowFactory(FindView);
    expect(wrapper.html()).not.toBe('');
  });
  it('matches the snapshot', async () => {
    const wrapper: Wrapper<Vue> = shallowFactory(FindView);
    // sets properties 'users', 'loans' and 'books' to true to allow their respective components to render
    await wrapper.setData({
      users: true,
      loans: true,
      books: true
    });
    // checks if the components rendered as stubs by comparing them to snapshot
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('requisite v-chips exist with correct content', () => {
    const wrapper: Wrapper<Vue> = shallowFactory(FindView);
    const chipStubs = wrapper.findAll('v-chip-stub');
    for (let i:number = 0; i < chipStubs.length; i++) {
      // Uses RegEx to chech if Users Loans or Books is included in text content of v-chip-stub
      if (/Users|Loans|Books/.test(chipStubs.at(i).text())) {
        expect(true).toBeTruthy();
      } else {
        expect(true).toBeFalsy();
      }
    }
  });
});
