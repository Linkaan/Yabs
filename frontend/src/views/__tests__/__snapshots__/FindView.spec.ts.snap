// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindView.vue matches the snapshot 1`] = `
<v-container-stub tag="div">
  <v-row-stub tag="div">
    <v-col-stub tag="div">
      <v-card-stub loaderheight="4" tag="div">
        <v-chip-group-stub activeclass="v-slide-item--active" mandatory="true" nexticon="$next" mobilebreakpoint="1264" previcon="$prev">
          <v-chip-stub color="success" activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" outlined="true" class="ma-2" value="true">
            <v-icon-stub left="">
              mdi-account-outline
            </v-icon-stub>
            Users
          </v-chip-stub>
          <v-chip-stub color="deep-purple" activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" outlined="true" class="ma-2" value="true">
            <v-icon-stub left="">
              mdi-handshake
            </v-icon-stub>
            Loans
          </v-chip-stub>
          <v-chip-stub color="light-blue" activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" outlined="true" class="ma-2" value="true">
            <v-icon-stub left="">
              mdi-library
            </v-icon-stub>
            Books
          </v-chip-stub>
        </v-chip-group-stub>
        <loanlistcomponent-stub items-per-page="15"></loanlistcomponent-stub>
        <booklistcomponent-stub use-actions="true" items-per-page="15"></booklistcomponent-stub>
        <userlistcomponent-stub items-per-page="15"></userlistcomponent-stub>
      </v-card-stub>
    </v-col-stub>
  </v-row-stub>
</v-container-stub>
`;
